commons/src/main/kotlin/com/github/quillraven/commons/audio/QueueAudioService.kt

```diff
@@ -0,0 +1,190 @@
+   * Making a call to [update] will play all pending requests.
+   */
+  override fun playSound(soundFilePath: String, volume: Float) {
+    if (soundRequests.size >= maxSimultaneousSounds) {
+      LOG.info { "Reached maximum of '$maxSimultaneousSounds' simultaneous sounds. Request is ignored" }
+      return
+    }
+
+    val request = soundRequests[soundFilePath]
+    if (request != null) {
+      // sound already queued -> set volume to maximum of both requests
+      request.volume = max(request.volume, volume).coerceIn(0f, 1f)
+    } else {
+      // sound not queued yet
+      soundRequests[soundFilePath] = SOUND_REQUEST_POOL.obtain().apply {
+        this.volume = volume
+        this.filePath = soundFilePath
+      }
+    }
+  }
+
+  /**
+   * Loads and plays a new [Music] instance for [musicFilePath] with the given [volume].
+   * If [loop] is true then the music will start again when finished.
+   *
+   * If there is already a music playing then it gets stopped and unloaded from the [assetStorage].
+   */
+  override fun playMusic(musicFilePath: String, volume: Float, loop: Boolean) {
+    if (currentMusicFilePath == musicFilePath) {
+      // trying to play the same music again -> ignore it
+      LOG.debug { "Music '$musicFilePath' is already playing" }
+      return
+    }
+
+    val oldMusic = currentMusic
+    if (oldMusic != null) {
+      // stop and unload current music
+      LOG.debug { "Stop music '$currentMusicFilePath'" }
+      oldMusic.stop()
+      KtxAsync.launch {
+        assetStorage.unload<Music>(currentMusicFilePath)
+      }
+    }
+
+    // verify music file path
+    if (!assetStorage.fileResolver.resolve(musicFilePath).exists()) {
+      LOG.error { "Music file '$musicFilePath' does not exist!" }
+      return
+    }
+
+    // load and play new music
+    LOG.debug { "Play music '$musicFilePath'" }
+    currentMusicFilePath = musicFilePath
+    currentMusic = assetStorage.loadSync<Music>(musicFilePath).apply {
+      this.isLooping = loop
+      this.volume = volume.coerceIn(0f, 1f)
+      play()
+    }
+  }
+
+  /**
+   * Pause the current active [Music].
+   */
+  override fun pauseMusic() {
+    currentMusic?.pause()
+  }
+
+  /**
+   * Resume the current active [Music].
+   */
+  override fun resumeMusic() {
+    currentMusic?.play()
+  }
+
+  /**
+   * Stop the current active [Music].
+   */
+  override fun stopMusic() {
+    currentMusic?.stop()
+  }
+
+  /**
+   * Processes all queued [sound requests][SoundRequest] of [playSound] and clears the queue.
+   * Will load the [Sound] via the [assetStorage] if needed and adds the [Sound] to an internal
+   * cache for faster access in the future.
+   *
+   * If the cache reaches the [maxCachedSounds] size then the [Sound] instances of the cache get unloaded
+   * from the [assetStorage] and the cache gets cleared.
+   */
+  override fun update() {
+    if (!soundRequests.isEmpty) {
+      soundRequests.iterate { _, request, iterator ->
```

Comment:
Don't do this. Pre-load the sounds and play the loaded files on demand. You're adding significant overhead by reloading the files each time you want to play a sound.

Reply:
I think you have a point for sound effects because even if I have 100s of them, the memory is most likely <10MB and noone cares if it gets loaded at startup.

For music it depends in my opinion. In my game it will not make a difference and I can also load it at startup or map change. I am thinking of adding that to the mapservice and in case someone defines a music property for the map then it will be played automatically.

I will think about that tomorrow :D

Also, I am not sure if the soundcache is even needed because assetStorage already provides that functionality? I also need to check that tomorrow again.

Reply:
Yeah, `AssetStorage` can store all of your sounds.

Reply:
As for the music files, it depends how many there are in the game, but normally I'd still at least preload a couple of themes (e.g. anything that has to do with the menu and the first few levels), and then gradually unload and load themes as the player progresses in the game. This load-on-demand approach will likely cause noticeable delays on mobile, especially since you're using synchronous loading.

Reply:
one question: the unloading of assetStorage is asynchronously, right? So that should not have an impact on the player experience?
Would it maybe make sense to keep the recent X music tracks in the service and if a music track falls out of the recent X tracks then it gets unloaded? That way the memory consumption will not grow endlessly. Or should this really be something that the user himself takes care of and I should always assume that it is handled by the user.
I would then still add a check that the Sound/Music is loaded by the assetStorage and if not log an error OR load it then synchronously.

Reply:
Unloading is asynchronous, but not executed in parallel unless you launch the coroutine in the `asyncContext`. Keep in mind that some assets (e.g. textures) might be necessary to dispose of on the rendering thread.

>  Would it maybe make sense to keep the recent X music tracks in the service and if a music track falls out of the recent X tracks then it gets unloaded? That way the memory consumption will not grow endlessly. Or should this really be something that the user himself takes care of and I should always assume that it is handled by the user.

I think this should be configurable by the user. You might try both approaches at once - allow the users to preload some music files that are never unloaded (e.g. menu theme), and keep the others in a cache with a fixed size.